iOS/TvShows/TvShows/Common/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,164 @@
+                    return
+                }
+                
+                guard let result: ResultType = self.parse(data: data) else {
+                    single(.failure(ApiError.noData))
+                    return
+                }
+
+                single(.success(result))
+            }
+            .resume()
+            
+            return Disposables.create()
+        }
+    }
+
+    private func execute<ParamsType: Encodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> Completable {
+        .create { [weak self] observer  in
+            guard let self = self else { return Disposables.create() }
+
+            let request = self.buildRequest(path: path, method: method, parameters: parameters)
+            self.urlSession.dataTask(with: request) { [weak self] _, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    observer(.error(ApiError.badRequest))
+                    return
+                }
+
+                if let error = self.mapToApiError(status: statusCode) {
+                    observer(.error(error))
+                    return
+                }
+
+                observer(.completed)
+
+            }
+            .resume()
+
+            return Disposables.create()
+        }
+    }
+
+    private func buildRequest<ParamsType: Encodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> URLRequest {
+        var components = URLComponents(string: "\(path)")!
+
+        if
+            method == .get || method == .delete,
+            let parameters = parameters as? [String: String]
+        {
+            components.queryItems = parameters.map(URLQueryItem.init)
+        }
+
+        var urlRequest = URLRequest(url: components.url!)
+        urlRequest.httpMethod = method.rawValue
+        urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
+
+        if
+            let parameters = parameters,
+            method == .post || method == .patch
+        {
+            urlRequest.httpBody = try? JSONSerialization.data(withJSONObject: parameters, options: .prettyPrinted)
+        }
+
+        return urlRequest
+    }
+
+    private func parse<ResultType: Decodable>(data: Data?) -> ResultType? {
+        guard let data = data else {
+            print("Data does not exist...")
+            return nil
+        }
+
+        do {
+            let dataString = String(decoding: data, as: UTF8.self)
+            print("Got data: \(dataString)")
+            return try JSONDecoder().decode(ResultType.self, from: data)
+        } catch {
+            print("Unexpected error on decoding data to \(ResultType.self)! (\(error)")
+            return nil
+        }
```

Comment:
This should be a throwing function since nil at the start should produce `noData` error and the catch block should produce some parse failure error.